Add docker-build-push-github action.

AlejoAsd opened the pull request:
This PR adds a new action that builds and pushes docker images to the Github registry. It can be run for branches, tags and pull requests.

AlejoAsd committed:
Add docker-build-push-github action.

AlejoAsd committed:
Run docker-build-push-github for tags.

AlejoAsd committed:
Make docker-build-push-github action run for the default branch.

AlejoAsd committed:
Allow running docker-build-push-github.yaml as workflow.

marcoshuck commented on .github/workflows/docker-build-push-github.yaml:
```diff
@@ -1,14 +1,19 @@
 # Based on https://docs.github.com/en/actions/publishing-packages/publishing-docker-images#publishing-images-to-github-packages
 
-name: Build a Docker image and publish it to the Github registry.
+name: Build Docker image and publish it to the Github registry.
```
```suggestion
name: Package OCI
```

marcoshuck replied:
I believe that names should be kept short to correctly display them in the Actions tab. As an alternative, what if we use `Build and publish OCIs` or something between those lines?

AlejoAsd committed:
Remove push triggers from docker-build-push-github.yaml.

Add image branch environment variable.
Add image branch to docker image based if not main/master branch.

AlejoAsd committed:
Simplify docker-build-push-github.yml branch append logic.

AlejoAsd committed:
Attempt to fix docker-build-push-github action branch name syntax error.

AlejoAsd committed:
Simplify branch name logic.

AlejoAsd committed:
The good stuff.

AlejoAsd committed:
Main and master. Change to single quotes.

AlejoAsd committed:
Use format for image branch concatenation.

AlejoAsd committed:
Remove image branch.

AlejoAsd committed:
Add env var echo.

AlejoAsd committed:
Name env var inspect step.

marcoshuck commented on .github/workflows/docker-build-push-github.yaml:
```diff
@@ -0,0 +1,50 @@
+          password: ${{ secrets.GITHUB_TOKEN }}
+          
+      - name: Inspect env vars
+        run: "echo REGISTRY: ${{ env.REGISTRY }}\nIMAGE_NAME: ${{ env.IMAGE_NAME }}"
```
```suggestion
        run: |
          echo REGISTRY: ${{ env.REGISTRY }}
          echo IMAGE_NAME: ${{ env.IMAGE_NAME }}
```

marcoshuck replied:
Not sure if the env vars will be printed this way, but there should be some mechanism to read them anyways.

AlejoAsd committed:
Explicitly use sh to display env vars.

AlejoAsd committed:
Display all env vars.

AlejoAsd committed:
Commit everything else.

AlejoAsd committed:
Restore.

AlejoAsd committed:
Convert slashes to dashes in image tag.

AlejoAsd committed:
Update environment variables.

AlejoAsd committed:
Add key to env expression.

AlejoAsd committed:
Update .github/workflows/docker-build-push-github.yaml

Co-authored-by: Marcos Huck <marcoshuck@users.noreply.github.com>

AlejoAsd committed:
Improve image tag definition.

AlejoAsd committed:
Uncomment steps.

AlejoAsd committed:
Test default branch.

AlejoAsd committed:
Add another default branch ref

AlejoAsd committed:
Use the Github event default branch to determine tag.

marcoshuck commented on .github/workflows/docker-build-push-github.yaml:
```diff
@@ -0,0 +1,61 @@
+        run: |
+          if [ "${IMAGE_TAG}" == "${{ github.event.repository.default_branch }}" ]
+          then
+            IMAGE_TAG=''
```
```suggestion
            IMAGE_TAG='latest'
```

marcoshuck replied:
What if we're explicit about it?

AlejoAsd replied:
Thanks! That's a good change.

marcoshuck commented on .github/workflows/docker-build-push-github.yaml:
```diff
@@ -0,0 +1,61 @@
+      packages: write
+
+    steps:
+      - name: Modify env var
```
There's an action that already takes care of this, and it's the one used below called `metadata`: https://github.com/docker/metadata-action#branch

AlejoAsd replied:
The metadata defines the value, but this step changes the image tag so that it doesn't contain any invalid characters like slashes. You can think of this step as pre-processing for the metadata step.

marcoshuck replied:
The action already does that , it's shown in the link referenced above.

marcoshuck replied:
| Event               | Ref                           | Output           |
|---------------------|-------------------------------|------------------|
| `pull_request`      | `refs/pull/2/merge`           | `pr-2`           |
| `push`              | `refs/heads/master`           | `master`         |
| `push`              | `refs/heads/my/branch`        | `my-branch`      |
| `push tag`          | `refs/tags/v1.2.3`            | `v1.2.3`         |
| `push tag`          | `refs/tags/v2.0.8-beta.67`    | `v2.0.8-beta.67` |
| `workflow_dispatch` | `refs/heads/master`           | `master`         |

marcoshuck replied:
Look how `my/branch` becomes `my-branch`.

AlejoAsd replied:
Oh, nice. I looked at the docs for that action and evidently missed that. This is actually great, thanks.

AlejoAsd replied:
This step has been removed in favor of `metadata-action` configuration options. Thanks for bringing this up.

AlejoAsd committed:
Use latest as image tag for default branch.

Co-authored-by: Marcos Huck <marcoshuck@users.noreply.github.com>
Signed-off-by: Alejo Carballude <alejocarballude@gmail.com>

AlejoAsd committed:
Only alter image tag for default branch.

AlejoAsd committed:
Use docker metadata action to define tags.

AlejoAsd committed:
Fix handlebars syntax error.

AlejoAsd committed:
Try with exclamation mark instead of not.

AlejoAsd committed:
Use Github Actions expression to determine branch tags.

marcoshuck reviewed (commented):
Looking good, can we test it out by consuming it from another repository?

NOTE: It might make sense to pass the `Dockerfile` as an input in the `workflow_call`.

```yaml
name: Package OCI
on:
  push:
    branches: [ $default-branch ]
  pull_request:

jobs:
  go:
    uses: gazebo-web/.github/.github/workflows/format-go.yaml@<ref>
    with:
      dockerfile: <path>
```

AlejoAsd committed:
Generate docker images for tags.

AlejoAsd commented:
> 

It is being tested in the [prometheus-json-exporter](https://github.com/gazebo-web/prometheus-json-exporter/blob/feature/ci/.github/workflows/docker.yml) repo.

marcoshuck reviewed (approved):
It seems to be working as expected. Approving!

AlejoAsd committed:
Add docker context and filepath inputs to workflow call.

AlejoAsd commented:
> NOTE: It might make sense to pass the Dockerfile as an input in the workflow_call.

These options have been added. I still need to test them.

marcoshuck commented on .github/workflows/docker-build-push-github.yaml:
```diff
@@ -7,6 +7,15 @@ name: Package OCI
   # This entry allows calling this workflow from other workflows
   workflow_call:
+    inputs:
+      docker-build-context:
```
This is redundant. What if we rename it to: `context`.

AlejoAsd replied:
Similar to my other comment, I find the `docker` prefix inoffensive. We can drop the `build` word from it.

marcoshuck replied:
Let's use `docker-context` then!

marcoshuck commented on .github/workflows/docker-build-push-github.yaml:
```diff
@@ -7,6 +7,15 @@ name: Package OCI
+        description: 'Build context is the PATH or URL to build from. '
+        type: string
+        default: .
+      docker-file-path:
```
I know you like to be verbose, and it's usually a right call, but I fail to see the value on doing that for this particular case. Why don't we call this `path` instead?

AlejoAsd replied:
I'm probably future-proofing for no good reason, but I find the `docker` prefix inoffensive and `file` is actually the parameter used by the `docker/build-push-action` action, which I didn't find intuitive enough, which is why I added the `path` suffix.

I can drop the `path` suffix if you want.

marcoshuck replied:
Let's use `dockerfile` or `filepath`, whatever you think is best.

AlejoAsd committed:
Temporarily remove workflow inputs.

AlejoAsd committed:
Remove with options from build-push-action.

AlejoAsd committed:
Remove with options.

AlejoAsd committed:
Remove branch ref enable parameter.

AlejoAsd committed:
Add docker inputs to docker-build-push-github action.

AlejoAsd committed:
Rename docker input. Add context to Dockerfile.

marcoshuck committed:
Merge branch 'main' into feature/docker-build-push
